sqlmesh/core/engine_adapter/base.py

```diff
@@ -67,6 +68,32 @@ def is_insert_overwrite(self) -> bool:
     def is_replace_where(self) -> bool:
         return self == InsertOverwriteStrategy.REPLACE_WHERE
 
+    @property
+    def requires_condition(self) -> bool:
+        return self.is_replace_where or self.is_delete_insert
+
+
+@dataclass
+class SourceQuery:
+    generator: t.Iterator[Query]
+    columns_to_types: t.Optional[t.Dict[str, exp.DataType]] = None
+    batched: bool = False
+    is_from_df: bool = False
+
+    @property
+    def single_query(self) -> t.Iterator[Query]:
+        assert not self.batched
+        yield from self.generator
+
+    def ensure_columns_to_types(
```

Comment:
I don't think I like this method. I'm not sure why is it a part of `SourceQuery` to begin with. I'm also not sure why does mutate it. There are quite a few levels of indirection here, while in actuality callers would've been better off just resolving it in place, ie.:
```
columns_to_types = source_query.columns_to_types or self.columns(table_name)
```

Reply:
Yeah I originally had that but found that I was putting it in a lot of places so I wanted to capture that logic and also cache the result.

Reply:
The caching was there previously anyway. We'd fetch columns and then pass it as an argument further  downstream. You essentially traded:
```
columns_to_types = source_query.columns_to_types or self.columns(table_name)
```
for
```
source_query. ensure_columns_to_types(table_name, self.columns)
```
So not sure whether we achieved anything here

Reply:
I went ahead and reverted back to how it was (which matches your top example).